### DISCUSSION TITLE
Herring data in comlandr

### DISCUSSION THREAD

**Discussion by sgaichas** [NOAA-EDAB/comlandr#67] [General]
The [`comlandr::get_comland_data`](https://github.com/NOAA-EDAB/comlandr/blob/dev/R/get_comland_data.R) **on the `dev` branch** uses several functions to build a commercial landings dataset.

In the first step, [`comlandr::get_comland_raw_data`](https://github.com/NOAA-EDAB/comlandr/blob/dev/R/get_comland_raw_data.R) pulls from the Stockeff table `stockeff.mv_cf_landings`, adds some fields based on other fields (small/large mesh), and optionally uses landed weight for shellfish and removes fish parts. It ensures that numbers are numeric, converts pounds to metric tons, identifies catch as US catch. A `comland` object is returned. 

Historically, not all species landings were contained in the same table. Therefore herring landings (collected by the State of Maine) are added separately.  

In the second step, [`comlandr::get_herring_data`](https://github.com/NOAA-EDAB/comlandr/blob/dev/R/get_herring_data.R), herring landings are pulled from a separate table, `NEFSC_GARFO.maine_herring_catch`. Numbers are converted to character, **kept and discarded tons are CORRECTED: NOT summed to catch**, fields are added for market code, tonnage class, and **utlization code**. Price is pulled from a subset of the above `comland` object by dividing SPPVALUE by SPPLIVMT by year and month. *So there is herring data in `comland` already.* 

At the end, herring CORRECTED: Landings ONLY from this function *replaces* all herring data in the original `comland` dataset that were pulled using the `get_comland_data` function.

The issue is that there are more rows in the `comland` object before adding the herring data than after, suggesting a loss of data. We are unclear on whether stockeff has already added herring in and need to check.

**Comment by sgaichas** [comment]
First step: save output of `get_comland_raw_data`, cut down to herring only, and compare with interim output of `get_herring_data`: `herring.catch`

Below we are comparing KEPT catch

![image](https://github.com/user-attachments/assets/f5db7af0-2fde-46b0-96d0-42e395744c57)

**Reply by andybeet** [reply]
so since about 2010 they seem almost identical! And the difference post 1980 is small. Do you think the difference is foreign catch? [NAFO by species](https://noaa-edab.github.io/comlandr/articles/ForeignCatchCompare21A21B.html#catch-by-species-1)

**Reply by andybeet** [reply]
Maybe the difference in not due to foreign landings. So this is the Herring data from NAFO (US landings excluded). Remarkably similar. I wonder if all of the Herring processing boats are considered foreign?

``` r
NEUSDivs <- c(47, 51:56, 61:63)
nafo <- comlandr::get_foreign_data()
#> Reading file: nafo-21b-60-69.zip
#> Reading file: nafo-21b-70-79.zip
#> Reading file: nafo-21b-80-89.zip
#> Reading file: nafo-21b-90-99.zip
#> Reading file: nafo-21b-2000-09.zip
#> Reading file: nafo-21b-2010-19.zip
#> Reading file: nafo-21b-2020-22.zip
nafo |>
  dplyr::filter(Code == 202,Divcode %in% NEUSDivs) |> 
  dplyr::group_by(Year) |> 
  dplyr::summarise(land = sum(SPPLIVMT)) |>
  ggplot2::ggplot() + ggplot2::geom_line(ggplot2::aes(x=Year,y=land))
```

![](https://i.imgur.com/SPWEr9I.png)

<sup>Created on 2024-09-04 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by sgaichas** [reply]
Do the NAFO landings get put back in for herring at a later `comlandr` step? Maybe they shouldn't if they are already in the ME dataset? I haven't looked that far downstream yet.

**Reply by andybeet** [reply]
Herring is removed from NAFO before being merged into comlandr pull https://github.com/NOAA-EDAB/comlandr/blob/master/R/process_foreign_data.R#L115

**Reply by sgaichas** [reply]
The herring pull from ME data has a CATEGORY field, which does suggest that NAFO is already added into that dataset. This compares the ME dataset "KEPT"  in the black line above as a stacked bar with the blue line that is the raw comland pull:

![image](https://github.com/user-attachments/assets/4b37146d-7225-4d89-8b67-1fff894653cd)

**Reply by andybeet** [reply]
This looks good right? ME herring data is complete.

**Comment by sgaichas** [comment]
The ME herring dataset includes KEPT and DISC fields that we currently sum in `comlandr::get_herring_data`. Discards are only recorded from 1996 on and have never exceeded 0.8% of the total catch.

However, summing the landings and discards here for herring is inconsistent with how we treat all other landings and discards. I *think* we only pull landings data for everything else, then the plan is to use the observer data to get kept/discard rates for the observed portion of the landings and apply those rates to the full landings. 

I *think* (?) this herring total catch will be treated as landings in subsequent steps to estimate discard, which is incorrect (although the proportion of discard is very small).

**Reply by andybeet** [reply]
I think you are right, the comlandr will estimate discards from this catch (landings + discards). I dont see an exclusion for herring

**Reply by sgaichas** [reply]
I was wrong about this, discards are dropped from the herring object. They are not summed with landings. `KEPTMT` is renamed `SPPLIVMT`. `DISCMT` is set to NULL. So herring should be consistent with all other species going forward into the discard estimation. 

We should consider whether we want to keep the DISCMT that comes from the ME herring data though.

**Comment by sgaichas** [comment]
See [issue](https://github.com/NOAA-EDAB/comlandr/issues/74) about documentation: no `SPPVALUE` prior to 1982?

There is code in `get_herring_data` that calculates herring price from the comlandr raw pull herring records as sum `SPPVALUE `/ sum `SPPLIVMT` over `YEAR` and `MONTH`.

The code uses an average price over all months from 1964 and applies that to all records before 1964. 

This price is then used to estimate `SPPVALUE` from `SPPLIVMT` in the herring data pulled from Maine (which has official landings and discards, but no value field).

This code makes me think we used to have `SPPVALUE` prior to 1982 in the comland raw pull. 

I remember some discussion that StockEff does not include value information from an older database. 

I think we don't calculate revenue or other `SPPVALUE` based indicators prior to 1982, but it might still be good to document the different lengths of `SPPLIVMT` and `SPPVALUE` time series for all species.

**Reply by andybeet** [reply]
You are right. There is SPPVALUE information prior to 1982 in the WOLAN tables but it is not available via STOCKEFF since there were "concerns as to the validity of the data. We have no way of validating it and had heard a few years ago that this field was not populated properly " 

When we used to pull from the WOLAN tables directly we did use the SPPVALUE.

**Comment by sgaichas** [comment]
Herring utilization codes from the comland_raw pull are used to proportion the herring landings from ME because this field is not in the ME dataset.

Codes are 

0=food fish or unknown; 
2=aquaculture; 
3=canned pet food (1984+); 
4=Biomedical (2002+); 
5=animal food (1984+); 
7=bait; 
8=industrial, other (2002+); 
9=industrial, reduction.

I think we split out "seafood" landings using these codes, 0 being seafood and all others not?

Here is e comparison of UTILCD (aggregated to 0 and all others=1) in comland_raw for herring with the total ME herring landings

![image](https://github.com/user-attachments/assets/aa45e74b-2be3-418a-a6d8-59cc942ab0bd)

**Reply by andybeet** [reply]
For the `comdat` dataset you are right, we use UTILCD = 0 only (except for menhaden where we use 7 & 9). In the herring pull we assume all catch is seafood and set UTILCD = 0. I see your point in the plot above. What are thinking we'd do to assign some of this catch to non-seafood?

**Reply by sgaichas** [reply]
The assignment UTILCD = 0  in `get_herring_data` is temporary to create the column in the ME herring dataset, later in the code the UTILCD from `comland_raw` is applied to the ME herring dataset using proportions by weight in the `comland_raw` dataset. So the output herring has codes 0, as well as 9, 7, 5, 3, and 2. I was just checking to see how things lined up. Lacking other data, I don't think we have any choice but to use the comland pull to estimate UTILCD in the full herring dataset.

**Reply by andybeet** [reply]
ah, i see! Yes, this makes sense.

**Comment by andybeet** [comment]
I also think that we should refactor `get_herring_data` to be complely independent and a stand alone function. Now it requires `comland` data as an argument which is unnecessary since the "cleaning" of the herring for comlandr purposes, can be done in another function or in the main body of `get_comland_data`. Then we can use the `get_herring data` in other projects without having to pull all of `Stockeff` first

**Reply by sgaichas** [reply]
So it would be just the pull from the GARFO table and fixing columns to be numeric? That seems reasonable. 

The herring data from comland is still needed to get prices and utilization codes and apply them to the herring ME dataset as noted above before herring from ME can be subbed into the comland dataset. But this could happen separately (`ME_herring_2comland` or something)

I could try separating them in the sarah_herring branch

**Reply by andybeet** [reply]
It's up to you. I wasn't suggesting we do this right now. We might want to wait if everything is doing what we want now. Was just thinking of cleaning, rerganizing for better use. Not urgent. And we might want to call it `process_herring_data`. This has the same implications then as the `process_foreign_data`. After seeing the assignment of UTILCD to herring based on `comlandr` data we might want to wait and think about this

**Reply by sgaichas** [reply]
I like both the suggested name and waiting till we clean up everything else.

**Comment by sgaichas** [comment]
Last sanity check, is our code getting the same data as are input into the assessment?

This compares the ME herring data from our pull that gets subbed into `comland` with fleet_1 (US) catch currently in the assessment. 

Looks pretty good although not an exact match. I can follow up with the assessment team

![image](https://github.com/user-attachments/assets/933e3816-c992-49b1-919a-e9a94a17ffc0)

**Reply by andybeet** [reply]
assessment is catch or landings? But looks good

**Reply by sgaichas** [reply]
I think it is catch but need to check. I plotted our data both ways and you can't see the difference because the discards are so small relative to the landings.
